### PR TITLE
[notations] Warn about incompatible prefixes

### PR DESCRIPTION
**Pull request by proux01**
Add a warning for notations with incompatible common prefixes, for instance
    
```Coq
Reserved Notation "#0 #1" (at level 30).
Reserved Notation "#0 #1 #2" (at level 40).
(* Warning: Notations "#0 #1" defined at level 30 and "#0 #1 #2" defined at level 40 have incompatible prefixes. One of them will likely not work. [prefix-incompatible-level,parsing,default] *)

Reserved Notation "#20 #21 x #3 y" (x at level 30, at level 50).
Reserved Notation "#20 #21 x #34" (x at level 40, at level 50).
(* Warning: Notations "#20 #21 _ #3 _" defined at level 50 with arguments constr at level 30 and "#20 #21 _ #34" defined at level 50 with arguments constr at level 40 have incompatible prefixes. One of them will likely not work. [prefix-incompatible-level,parsing,default] *)
```

- [x] Added / updated **test-suite**.

- [x] Added **changelog**.
- [x] Added / updated **documentation**.
  
  - [x] Documented any new / changed **user messages**.
  - [ ] ~Updated **documented syntax** by running `make doc_gram_rsts`.~

- [ ] ~Opened **overlay** pull requests.~

<!-- Pointers to relevant developer documentation:

Contributing guide: https://github.com/coq/coq/blob/master/CONTRIBUTING.md

Test-suite: https://github.com/coq/coq/blob/master/test-suite/README.md

Changelog: https://github.com/coq/coq/blob/master/doc/changelog/README.md

Building the doc: https://github.com/coq/coq/blob/master/doc/README.md
Sphinx: https://github.com/coq/coq/blob/master/doc/sphinx/README.rst
doc_gram: https://github.com/coq/coq/blob/master/doc/tools/docgram/README.md

Overlays: https://github.com/coq/coq/blob/master/dev/ci/user-overlays/README.md

### CODE REVIEW

**Comment by proux01** [comment]
@herbelin this is the first step toward the `at existing level` notation option we already discussed and that I'd like to add in a further PR.

**Comment by SkySkimmer** [comment]
Maybe the warning should explain what the consequences of incompatible prefixes are (I don't know myself)

**Comment by proux01** [comment]
Indeed, done, as well as the documentation.

File: doc/changelog/03-notations/19049-warn_notation_incompatible_prefix.rst [hunk 1]
```diff
@@ -0,0 +1,4 @@
+- **Added:**
+  a :warn:`notation-incompatible-prefix` warning
```
Comment (jfehrle) [inline doc/changelog/03-notations/19049-warn_notation_incompatible_prefix.rst:2]:
```suggestion
  :warn:`notation-incompatible-prefix` warning when two notation definitions have incompatible prefixes
```

File: doc/sphinx/user-extensions/syntax-extensions.rst [hunk 2]
```diff
@@ -215,6 +215,8 @@ definition is the following:
 More generally, it is required that notations are explicitly factorized on the
 left. See the next section for more about factorization.
 
+.. _NotationFactorization:
+
 Simple factorization rules
```
Comment (jfehrle) [inline doc/sphinx/user-extensions/syntax-extensions.rst:220]: This section doesn't explain why the user cares about factorization, so hard to follow.  Perhaps you could add a sentence or two to improve this? Maybe something like:
- distinct notations must be compatible
- in some cases, incompatible notations can be supported by factorizing them.
- "factorizing" means <add short description>  (it's a generic term but we have a very specific meaning in mind, described in the section)

IIUC the point is that if one Notation is a prefix of another, you need to factorize the common prefix so that both can be parsed correctly.

File: doc/sphinx/user-extensions/syntax-extensions.rst [hunk 3]
```diff
@@ -436,6 +438,13 @@ Reserving notations
    default by all subsequent interpretations of the corresponding
    notation. Individual interpretations can override the format.
 
+   .. warn:: Notations "a b" defined at level x and "a c" defined at level y have incompatible prefixes. One of them will likely not work.
+      :name: notation-incompatible-prefix
+
+      The levels of one of the notations should be adjusted to match
+      the other. See :ref:`factorization <NotationFactorization>` for
+      details.
```
Comment (jfehrle) [inline doc/sphinx/user-extensions/syntax-extensions.rst:446]: How about precisely describing when two notations are incompatible?

File: test-suite/output/notation_prefix_incompatible_level.out [hunk 4]
```diff
@@ -0,0 +1,9 @@
+File "./output/notation_prefix_incompatible_level.v", line 2, characters 0-43:
+Warning: Notations "#0 #1" defined at level 30 and "#0 #1 #2"
+defined at level 40 have incompatible prefixes.
```
Comment (jfehrle) [inline test-suite/output/notation_prefix_incompatible_level.out:3]:
```suggestion
File "./output/notation_prefix_incompatible_level.v", line 2, characters 0-43:
Warning: Notations "#0 #1" defined at level 30 (in File "foobar", line 1, 0-40) and "#0 #1 #2"
defined at level 40 have incompatible prefixes.
```

This message would be better if it gave the location of the first notation.  Otherwise the user has to search/scan for the definition, which could in be many places.

**Review by jfehrle — COMMENTED** [review]
wording and other suggestions

**Comment by jfehrle** [comment]
@SkySkimmer > Maybe the warning should explain what the consequences of incompatible prefixes are (I don't know myself)

IIUC the parser would only recognize one of the notations (and report syntax errors on the others), depending on the order in which the parser processes the rules defined for each notation (first one wins).  Our parser really doesn't do lookahead--it looks at the next symbol before making a parsing decision.  Once the first token of a rule is accepted, the parser is unable to switch to a different rule.  Not like LALR(k) parsers such as yacc and bison.  (There is a hard coded exception to this behavior for command names.)

File: test-suite/output/notation_prefix_incompatible_level.out (hunk 4 above)
Comment (proux01) [reply test-suite/output/notation_prefix_incompatible_level.out:3]: Yes, in fact we could need the locations of the two notations because the warning can also be raised when importing two incompatible modules.
However, this is clearly outside the scope of the current PR. IIRC there were already attempts by @ejgallego to locate things but this raised interesting questions and did not converge.

**Comment by SkySkimmer** [comment]
@jfehrle please don't guess
>There is a hard coded exception to this behavior for command names

is certainly not true so I don't trust the rest of your post.

File: test-suite/output/notation_prefix_incompatible_level.out (hunk 4 above)
Comment (jfehrle) [reply test-suite/output/notation_prefix_incompatible_level.out:3]: Surely it can't be that hard to add a Loc.t to the notation data structure, but it's your call.

Comment (proux01) [reply test-suite/output/notation_prefix_incompatible_level.out:3]: Well, prior experiences show it is definitely non trivial.

**Comment by proux01** [comment]
@coqbot run full ci

**Comment by jfehrle** [comment]
@SkySkimmer The exceptional code is `Grammar.parser_of_token_list`, which does multisymbol lookahead among prefixes containing only terminal symbols.  Most relevant to command names, option names and tactic names and not used for most other nonterminals.

File: doc/sphinx/user-extensions/syntax-extensions.rst [hunk 5]
```diff
@@ -215,15 +215,22 @@ definition is the following:
...
 Coq extensible parsing is performed by *Camlp5* which is essentially a LL1
 parser: it decides which notation to parse by looking at tokens from left to right.
 Hence, some care has to be taken not to hide already existing rules by new
-rules. Some simple left factorization work has to be done. Here is an example.
+rules. Indeed notations with a common prefix setting different levels could
+interfere with one another. For instance, a notation ``x < y`` with ``x``
+and ``y`` at level 69 would be broken by a further rule that would put
+``y`` at another level, like ``x < y < z`` with ``x`` at level 69 and ``y``
+at level 200. To avoid such issues, one should left factorize rules, that is ensure
+that common prefixes set the samel levels.
```
Comment (jfehrle) [inline doc/sphinx/user-extensions/syntax-extensions.rst:231]:
```suggestion
rules. Indeed, notations with a common prefix setting different levels could
interfere with one another. For instance, a notation ``x < y`` with ``x``
and ``y`` at level 69 would be broken by a further rule that would put
``y`` at another level, like ``x < y < z`` with ``x`` at level 69 and ``y``
at level 200. To avoid such issues, one should left factorize rules, that is ensure
that common prefixes set the same levels.
```

File: doc/sphinx/user-extensions/syntax-extensions.rst [hunk 6]
```diff
@@ -215,15 +215,22 @@ definition is the following:
...
 Coq extensible parsing is performed by *Camlp5* which is essentially a LL1
 parser: it decides which notation to parse by looking at tokens from left to right.
 Hence, some care has to be taken not to hide already existing rules by new
-rules. Some simple left factorization work has to be done. Here is an example.
+rules. Indeed, notations with a common prefix setting different levels could
+interfere with one another. For instance, a notation ``x < y`` with ``x``
+and ``y`` at level 69 would be broken by a further rule that would put
+``y`` at another level, like ``x < y < z`` with ``x`` at level 69 and ``y``
+at level 200. To avoid such issues, one should left factorize rules, that is ensure
+that common prefixes set the same levels.
```
Comment (jfehrle) [inline doc/sphinx/user-extensions/syntax-extensions.rst:231]:
```suggestion
rules. Indeed, notations with a common prefix but different levels can
interfere with one another, causing some of the them to parse incorrectly.  For instance, a notation ``x < y`` with ``x``
and ``y`` at level 69 would be broken by another rule that puts
``y`` at another level, like ``x < y < z`` with ``x`` at level 69 and ``y``
at level 200. To avoid such issues, you should left factorize rules, that is ensure
that common prefixes use the same levels.
```

Should it be "can interfere" or "will interfere"?

Comment (proux01) [reply doc/sphinx/user-extensions/syntax-extensions.rst:231]: Can interfere, since in some articular cases both notation can work fine.
What about "notations with a common prefix but different levels can
interfere with one another, causing some of them to not parse."? (since what actually happens is that one of the notations (sometimes both) becomes unusable)

Comment (jfehrle) [reply doc/sphinx/user-extensions/syntax-extensions.rst:231]: How about "not to work"?  or (better) "makes some of them unusable"?
(probably gives parse errors but the doc doesn't need to be that specific).

**Comment by herbelin** [comment]
Going to merge, after rebase and clean CI.

**Review by herbelin — APPROVED** [review]

**Comment by proux01** [comment]
@herbelin rebased, CI green

**Comment by herbelin** [comment]
@coqbot run full ci

**Comment by herbelin** [comment]
@coqbot merge now

**Comment by tchajed** [comment]
This error message produces spurious error messages in Iris. For example, in the proofmode notations (https://gitlab.mpi-sws.org/iris/iris/-/blob/0da37f400b47ff58f8b57003aed398f6d4a2eb41/iris/proofmode/notation.v), the line 28 notation is considered incompatible with the line 19 notation, but it does work for printing when `Δ = Enil`. You can see this by going to [tests/proofmode.v line 35](https://gitlab.mpi-sws.org/iris/iris/-/blob/0da37f400b47ff58f8b57003aed398f6d4a2eb41/tests/proofmode.v#L35) and you'll get a line `  --------------------------------------□` which is undesirable (`Δ = Enil` in this case). 

I found if I reverse these notations (so the "more specific" notation is first), the error message says Δ is set to level 200, and doing this manually does suppress the warning.

Can the error message handle this case and avoid giving a warning? Should we put `Δ at level 200` and is that really necessary?

**Comment by proux01** [comment]
As the message says, the non matching levels *may* cause things to not parse. Sometimes things do work, for instance due to the recovery mechanism of camlp5, that we plan to remove in the near future. So you can silence the warning if you are certain it's spurious (for instance with the attribute `#[warning="-warning-name"]`) but it's not necessarily a good idea.

**Comment by herbelin** [comment]
The detection could still be improved, I think, no? It looks here as if an inner constr whose level is left unspecified (the `Δ` of the first notation) were not identified with level 200, even though it is eventually adjusted to 200 (as shown by `Print Grammar constr`).

**Comment by robbertkrebbers** [comment]
The notations in the Iris file that @tchajed mentions are `printing only`, so they should not be used for parsing. So perhaps the error message should simply only be used for notations that are *not* `printing only`.
